### PR TITLE
added IS_STR, IS_NUM, VALID to formula language

### PR DESCRIPTION
**Pull request by o-smirnov**
Adding the power spectrum branch to TRON strained at the limits of the parser a little, so I added these functions to make things more tractcable.

### CODE REVIEW

File: docs/source/fundamentals/substitutions.rst [hunk 1]
```diff
@@ -169,6 +169,15 @@ As we saw above, a parameter value starting with ``=`` invokes the formula parse
    
   * ``STRIPEXT(`` *path* ``)`` returns the path minus the extension. 
 
+  * ``IS_NUM(arg)`` true if the argument is a numeric type
+
+  * ``IS_STR(arg)`` true if the argument is a string type
+
+  * ``VALID(arg)`` true if the argument is valid, and evaluates to non-zero. This is a useful pattern when dealing 
+  with parameters of a mixed type (that can be e.g. strings or numbers). For example, ``recipe.a > 0`` would throw an 
+  error is ``a`` is a string, but ``VALID(recipe.a > 0)
```
Comment (JSKenyon) [inline docs/source/fundamentals/substitutions.rst]: This documentation looks incomplete/cut off.

**Review by JSKenyon — CHANGES_REQUESTED** [review]
Just one minor comment about the docs.

File: scabha/evaluator.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import os.path
 import fnmatch
 import pyparsing
+import numpy as np
```
Comment (JSKenyon) [inline scabha/evaluator.py:6]: Numpy needs to be added to to the dependencies - tests currently fail.

**Review by JSKenyon — DISMISSED** [review]

**Comment by JSKenyon** [comment]
3.12 tests are failing due to deprecated functionality. I would personally advocate for evicting 3.8 from the test matrix/setup. It has moved to EOL as of 2024-10-07 (see https://devguide.python.org/versions/).

**Comment by o-smirnov** [comment]
I'm not particularly attached to 3.8, but how does dropping it help the 3.12 issue?

**Comment by JSKenyon** [comment]
> I'm not particularly attached to 3.8, but how does dropping it help the 3.12 issue?

Managed to reproduce locally. The poetry installer tries to produce a uniform environment over all the Python versions. That results in an old numpy being installed to satisfy Python 3.8. I have pushed a fix - let us see if it works.

**Comment by o-smirnov** [comment]
@JSKenyon @SpheMakh please re-review.

**Review by JSKenyon — APPROVED** [review]

**Pull request merged**